server/db.py

```diff
@@ -334,3 +375,72 @@ def get_encrypted_sym_key(username: str, user_password, friend_username:str):
 
 
+#---DIRECT MESSAGE FUNCTIONS---#
+
+def concatenate_usernames(username1:str, username2:str)->str:
+    """Function that will trim usernames if they are too long and then concatenate them
+    alphabetically with an underscore between them. Returns the concatenated string."""
+    # sort usernames alphabetically to ensure consistent ordering
+    username1 = username1[:5]
+    username2 = username2[:5]
+    sorted_usernames = sorted([username1, username2])
+    # join sorted usernames with an underscore
+    concatenated_usernames = '_'.join(sorted_usernames)
+    return concatenated_usernames
+
+
+def generate_database_key_hash(username1:str, username2:str)->str:
+    """Function that will take the concatenated string of usernames and generate a reproducable hash
+    to serve as a database key."""
+    # concatenate usernames alphabetically
+    concatenated_usernames = concatenate_usernames(username1, username2)
+    # hash concatenated usernames using SHA-256 hash function
+    hash_object =hashlib.sha256(concatenated_usernames.encode())
+    key_hash_hex = hash_object.hexdigest()
+    return key_hash_hex
+
+def get_message_status(generated_database_key_hash:str):
+    """Function that will return the highest order number for a conversation. If no conversation exist it will return
+    None."""
+    status = DM_MESSAGES.get(generated_database_key_hash)
+    if status:
+        return max(map(int, status["conversation"].keys()))
+    return 
+
+def push_message_to_database(speaker:str, receiver:str, message:str):
+    """Function that will enter a new conversation in the database in none exists. If it does exist it will update the conversation
+    with the newest message."""
+    generated_database_key_hash = generate_database_key_hash(speaker, receiver)
+    
+    order_number = get_message_status(generated_database_key_hash)
+    if order_number:
+        order_number+=1
+        update = {
+            f"conversation.{order_number}" : {
+            "conversation.order_number.speaker" : speaker,
+            "conversation.order_number.text" : message,
+            "conversation.order_number.date_time" : str(datetime.datetime.utcnow())
+        }}
+        try:
+            DM_MESSAGES.update(update, generated_database_key_hash)
+        except Exception as error_message:
+            logging.exception(error_message)
+            return 
+    else:
+        order_number = 1
+    
+        update = {
+        "key": generated_database_key_hash,
+        "conversation": {
+            order_number: {
+                "speaker": speaker,
+                "text": message,
+                "date_time": str(datetime.datetime.utcnow())
+            }    }  }
+        try:
+            DM_MESSAGES.put(update)
+        except Exception as error_message:
+            logging.exception(error_message)
+            return
+
+update_thought_rating("222a8c6f-13a0-4f4c-8ae9-8f1e2dae6b8a")
```

Comment:
What is this? I'd put this in an environment variable if I were you.

Reply:
it is a dynamically generated hash based on the two usersnames. It will be used to easily retrieve conversations. If you know the username and the friends username the correct hash will be generated(they are sorted alphabetically and trimmed to 5 characters if longer than 5, before hash generation). It is not so much a security feature. It is more to easily store and retrieve the conversations. There will be one such hash per conversation between two people

Reply:
plus that update_thought_rating at the bottom needs to go, i was testing functionality :-P

Reply:
Understood